tools/reviewer/job/impl/CITask.java

```diff
@@ -0,0 +1,122 @@
+import com.google.startupos.tools.reviewer.ReviewerProtos.Repo;
+import com.google.startupos.tools.reviewer.job.tasks.Task;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class CITask extends FirestoreTaskBase implements Task {
+  private static FluentLogger log = FluentLogger.forEnclosingClass();
+
+  private final ReentrantLock lock = new ReentrantLock();
+
+  private FileUtils fileUtils;
+  private GitRepoFactory gitRepoFactory;
+
+  private static String CI_REQUESTS_PATH = "/reviewer/ci/requests";
+  private static String CI_RESPONSES_PATH = "/reviewer/ci/responses";
+
+  @Inject
+  public CITask(
+      FileUtils fileUtils,
+      GitRepoFactory gitRepoFactory,
+      FirestoreClientFactory firestoreClientFactory) {
+    this.fileUtils = fileUtils;
+    this.gitRepoFactory = gitRepoFactory;
+    this.firestoreClientFactory = firestoreClientFactory;
+  }
+
+  @Override
+  public Boolean shouldRun() {
+    return !lock.isLocked();
+  }
+
+  @Override
+  public void run() {
+    if (lock.tryLock()) {
+      try {
+        initializeFirestoreClientIfNull();
+
+        log.atInfo().log("Running CI job (Firestore client @ %s)", firestoreClient);
+
+        CIRequest request =
+            (CIRequest) this.firestoreClient.popDocument(CI_REQUESTS_PATH, CIRequest.newBuilder());
+
+        if (request == null) {
+          // there is no request to process
+          return;
+        }
+
+        if (!fileUtils.folderExists("ci")) {
+          fileUtils.mkdirs("ci");
+        }
+
+        CIResponse.Builder responseBuilder = CIResponse.newBuilder().setRequest(request);
+
+        for (CIRequest.Target target : request.getTargetList()) {
+          Repo repo = target.getRepo();
+
+          String repoPath =
+              fileUtils.joinPaths(fileUtils.getCurrentWorkingDirectory(), "ci", repo.getId());
+
+          GitRepo gitRepo = this.gitRepoFactory.create(repoPath);
+
+          if (fileUtils.folderEmptyOrNotExists(repoPath)) {
+            gitRepo.cloneRepo(repo.getUrl(), repoPath);
+          } else {
+            gitRepo.pull();
+          }
+
+          gitRepo.switchBranch(target.getCommitId());
+
+          log.atInfo().log("Running reviewer-ci.sh in %s", repoPath);
+          CommandLine.CommandResult result =
+              CommandLine.runCommandForError(
+                  "/usr/bin/env bash " + fileUtils.joinPaths(repoPath, "reviewer-ci.sh"));
+
+          responseBuilder.addResults(
+              CIResponse.TargetResult.newBuilder()
+                  .setTarget(target)
+                  .setSuccess(result.exitValue == 0)
+                  // TODO: log truncation strategy?
+                  .setLog(result.stderr)
+                  .build());
+        }
+
+        firestoreClient.createDocument(CI_RESPONSES_PATH, responseBuilder.build());
+      } catch (IOException e) {
+        // TODO: store in Firestore?
```

Comment:
Yes, you can store the exception in the log.

Reply:
@oferb actually we don't need to store it in `CI` log -- `IOException` here is thrown by `folderEmptyOrNotExists` and this is most likely caused by server misconfiguration

Reply:
I think it would still be useful to see the log in the same way we use to view the CI failures.
Otherwise, we'd have to hunt it down in the server logs somehow.

Reply:
@oferb fixed in b22959393ec208c96c2923826ab2dedce7eff98b